### DISCUSSION TITLE
How to typehint an instance of a class that is going to implement multiple Protocols in a context?

### DISCUSSION THREAD

**Discussion by FeryET** [python/typing#971] [Q&A]
I have also asked this question [here](https://stackoverflow.com/q/70141151/8265036).

I have an object inside a context I want to use from both it's superclass methods, and given the function, use of its multiple protocol methods it has implemented. How can I typehint it in a way so mypy understands that?

**Comment by srittau** [comment]
@FeryET Could you give an example so it's easier to understand what you mean?

**Comment by Gobot1234** [comment]
Looking at the stack overflow post you should just use the Father type hint as that's how duck typing works, Father quacks like DummyProtocol as it defines do_something. If that's not what you mean, I need another example.

**Reply by FeryET** [reply]
> Looking at the stack overflow post you should just use the Father type hint as that's how duck typing works, Father quacks like DummyProtocol as it defines do_something. If that's not what you mean, I need another example.

So if I have multiple protocols, and want a combination of them in a certain context, what should I do? For example we have a baseclass Father, and multiple protocols, ProtocolA, ProtocolB, ProtocolC, and a some children of Father only implement some of these protocols.

**Reply by Gobot1234** [reply]
You need an intersection or to create a new protocol with all of the bases you need

**Reply by FeryET** [reply]
> You need an intersection or to create a new protocol with all of the bases you need

How can I define an intersection? Does python `Typing` have any intersection class? Because I couldn't find anything before posting this.

**Reply by Gobot1234** [reply]
Not currently no but there is a PEP for it, which is why I gave the second option of creating another protocol with all of the methods required

**Reply by a-reich** [reply]
Or you can create a new protocol class that just inherits from each of the previous protocols, right? 

class ProtoIntersection(ProtocolA, ProtocolB, ProtocolC, typing.Protocol): …

**Reply by Gobot1234** [reply]
Yes

**Reply by FeryET** [reply]
> Or you can create a new protocol class that just inherits from each of the previous protocols, right?
> 
> class ProtoIntersection(ProtocolA, ProtocolB, ProtocolC, typing.Protocol): …

That is a bit inconvenient tho, especially for something that is only used for static typing like Protocols. It reduces readability and transparency.

**Reply by Gobot1234** [reply]
Yes this is one of the proposed use cases for an intersection as you could just do `ProtocolA & ProtocolB & ProtocolC`
